### DISCUSSION TITLE
Working with C++ input for WarpX

### DISCUSSION THREAD

**Discussion by msaadeh0** [BLAST-WarpX/warpx#5062] [Installation] [answered]
Hi everyone,

I am new to WarpX and I want to write a code using the C++ classes, build it using CMAKE then run it. I managed to do all the steps but when I create my own CMake file to build my code, the building process fails because the file WarpXTargets.cmake is not available. I checked the repository online and I could not find here either. I am wondering what am I doing wrong. Any advice?

This is my CMakeLists.txt:

```cmake
cmake_minimum_required(VERSION 3.10)
project(WarpXTest)


set(CMAKE_CXX_STANDARD 14)


set(AMREX_DIR //home/USER/WarpX/build/_deps/fetchedamrex-build/lib/cmake/AMReX)
set(WARPX_DIR /home/USER/WarpX)  



find_package(AMReX REQUIRED PATHS ${AMREX_DIR})
find_package(WarpX REQUIRED PATHS ${WARPX_DIR})
find_package(HDF5 REQUIRED)


add_executable(WarpXTest main.cpp)


target_include_directories(WarpXTest PRIVATE ${WarpX_INCLUDE_DIRS} ${AMREX_INCLUDE_DIRS} ${HDF5_INCLUDE_DIRS})
target_link_libraries(WarpXTest PRIVATE WarpX::WarpX ${AMREX_LIBRARIES} ${HDF5_LIBRARIES} sz)
```

The error I receive is:
```
CMake Error at /home/USER/WarpX/WarpXConfig.cmake:55 (include):
  include could not find requested file:

    /home/USER/WarpX/WarpXTargets.cmake
Call Stack (most recent call first):
  CMakeLists.txt:14 (find_package)
```

Thanks

**Comment by ax3l** [comment, marked as answer]
Hi @msaadeh0,

While you can already use the library ABLASTR inside WarpX to build your own PIC code, e.g., see [ImpactX](https://github.com/ECP-WarpX/impactx) logic for this, WarpX does not yet fully support this. We are working on a C++ library install of WarpX in https://github.com/ECP-WarpX/WarpX/pull/5008 .

In the meantime, you could simply `add_subdirectory()` to the WarpX source directory and skip the `find_package`. This is called a "superbuild" and will do the same as you are trying to do with the snippet above, but without having WarpX pre-installed as binaries (instead, it will build it on the fly).

But more generally, which specific use case are you after? Maybe we can guide you to do something different or contribute to WarpX mainline. Or maybe, you can simply extend WarpX from Python, using our Python bindings.

**Reply by msaadeh0** [reply]
Thanks @ax3l 

I am trying to use WarpX as a library to build an electrostatic 3D simulation. There are examples on how to do that using PICMI, that was the first option but for some reason it did not work and to be honest Python is not my strong suit. I thought it might be more straightforward to use WarpX as a linked library to a C++ code. If there is an example of how to do that please let me know.

I will try the superbuild approach and I hope it will work. I will update the answer here on how it went

Thanks very much
